builtins: add repaired_descriptor builtin

postamar opened the pull request:
This commit adds a new crdb_internal.repaired_descriptor builtin
which makes use of the recently-added StripDanglingBackReferences method
on catalog.MutableDescriptor.

In conjunction with crdb_internal.unsafe_upsert_descriptor, this builtin
makes it possible to repair descriptors corrupted with dangling
back-references (back-references to descriptors or jobs which are
invalid) without needing to craft a descriptor surgery query for each
descriptor. It is therefore possible to repair this whole class of
corruptions in bulk.

This class of descriptor corruptions is particularly insidious in that
it is only caught by validation checks performed when writing new
versions of the descriptor (i.e. schema changes). This is deliberate, in
that we don't want this kind of corruption to interfere with DML queries
because those don't interact with back-references at all. On the other
hand, it defers detection of the corruption until a schema change or
a cluster upgrade is attempted.

Informs #104425.

Release note: None

Commit:
catalog: add StripDanglingBackReferences method

This commit adds this method to the MutableDescriptor interface.

Release note: None

Commit:
builtins: add repaired_descriptor builtin

This commit adds a new crdb_internal.repaired_descriptor builtin
which makes use of the recently-added StripDanglingBackReferences method
on catalog.MutableDescriptor.

In conjunction with crdb_internal.unsafe_upsert_descriptor, this builtin
makes it possible to repair descriptors corrupted with dangling
back-references (back-references to descriptors or jobs which are
invalid) without needing to craft a descriptor surgery query for each
descriptor. It is therefore possible to repair this whole class of
corruptions in bulk.

This class of descriptor corruptions is particularly insidious in that
it is only caught by validation checks performed when writing new
versions of the descriptor (i.e. schema changes). This is deliberate, in
that we don't want this kind of corruption to interfere with DML queries
because those don't interact with back-references at all. On the other
hand, it defers detection of the corruption until a schema change or
a cluster upgrade is attempted.

Informs #104425.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105746)
<!-- Reviewable:end -->

Commit:
catalog: remove GetReferencingDescriptorIDs from type desc

We don't like exposing the slices in the underlying protobuf.

Release note: None
